go-controller/pkg/node/node.go

```diff
@@ -540,29 +541,40 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
 	return err
 }
 
-func (n *OvnNode) WatchEndpoints() {
-	n.watchFactory.AddEndpointsHandler(cache.ResourceEventHandlerFuncs{
+func (n *OvnNode) WatchEndpointSlices() {
+	n.watchFactory.AddEndpointSliceHandler(cache.ResourceEventHandlerFuncs{
 		UpdateFunc: func(old, new interface{}) {
-			epNew := new.(*kapi.Endpoints)
-			epOld := old.(*kapi.Endpoints)
-			newEpAddressMap := buildEndpointAddressMap(epNew.Subsets)
-			for item := range buildEndpointAddressMap(epOld.Subsets) {
-				if _, ok := newEpAddressMap[item]; !ok {
-					err := deleteConntrack(item.ip, item.port, item.protocol)
-					if err != nil {
-						klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
+			newEndpointSlice := new.(*discovery.EndpointSlice)
+			oldEndpointSlice := old.(*discovery.EndpointSlice)
+			for _, port := range oldEndpointSlice.Ports {
+				for _, endpoint := range oldEndpointSlice.Endpoints {
+					for _, ip := range endpoint.Addresses {
+						if isEPSliceContainsEndpoint(newEndpointSlice, ip, *port.Port, *port.Protocol) {
+							continue
+						}
+						if *port.Protocol == kapi.ProtocolUDP || *port.Protocol == kapi.ProtocolSCTP {
+							err := deleteConntrack(ip, *port.Port, *port.Protocol)
+							if err != nil {
+								klog.Errorf("Failed to delete conntrack entry for %s: %v", ip, err)
+							}
+						}
 					}
 				}
 			}
 		},
 		DeleteFunc: func(obj interface{}) {
-			ep := obj.(*kapi.Endpoints)
-			for item := range buildEndpointAddressMap(ep.Subsets) {
-				err := deleteConntrack(item.ip, item.port, item.protocol)
-				if err != nil {
-					klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
+			endpointSlice := obj.(*discovery.EndpointSlice)
+			for _, port := range endpointSlice.Ports {
+				for _, endpoint := range endpointSlice.Endpoints {
+					for _, ip := range endpoint.Addresses {
+						if *port.Protocol == kapi.ProtocolUDP || *port.Protocol == kapi.ProtocolSCTP {
```

Comment:
Why do we need to manually delete the conntrack rules here @trozet Shouldn't they just manually timeout and be removed?

Reply:
UDP conntrack rules can take a while to time out because of the inherent statlessness in UDP..